### DISCUSSION TITLE
WGSL (Javascript) Inquiry

### DISCUSSION THREAD

**Discussion by TalDerei** [gpuweb/gpuweb#2384] [Q&A]
Is it possible to take existing javascript code, package it in compute shader blocks, and send that off to the GPU? Or does the Javascript need to be translated to WGSL inside the createShaderModule()? Thanks!

**Comment by Rezmason** [comment]
I'm not sure I understand your question, but I'd like to try.

There are no current plans for WebGPU to support JavaScript as a _shader language_, ie. the syntax that is compiled by `GPUDevice::createShaderModule` into GPU instructions. It is possible for you to write a JavaScript project that _generates_ WGSL, but I'm not sure what you would generate it from, unless you are transpiling shaders written in other existing shader languages.

For the first version of WebGPU, all the code you want the GPU to run in a single compute pipeline has to go into a single WGSL shader with a specific compute entry point for that pipeline. If you have multiple functions you want to run on the GPU, you can make each of them a compute entry point in your WGSL shader, create a shader module, and create a pipeline from it for each of your endpoints. You can then dispatch those pipelines in any order you want.

Did I answer your question?

_Edit: fixed a typo._

**Reply by Rezmason** [reply]
I should add that you can also create multiple shader modules, each from a different WGSL source, and create pipelines from those. And because the WGSL code starts as a string, you can share code between WGSL shaders by combining different strings before compiling them.

**Reply by TalDerei** [reply]
Thanks!! I have another question, how do I load an existing SPIRV kernel to run on webGPU?

**Reply by Rezmason** [reply]
I speak for myself, but at this moment, I don't recommend you do that.

WebGPU is still a work in progress. From my perspective as a recent beginner like you, you will want to write your shaders in WGSL for the time being. There might be community members who maintain libraries that convert higher level languages to WGSL, but converting SPIR-V to WGSL [turns out not to be trivial or practical](http://kvark.github.io/spirv/2021/05/01/spirv-horrors.html).

**Reply by Kangz** [reply]
[GPU.js](https://gpu.rocks/#/) has something to automatically translate Javascript to WebGL GLSL shaders so maybe a similar approach is possible for WebGPU / WGSL.

If you're making new code, WGSL should be the way to go but if you already have a SPIRV pipeline then you can use tools that translate SPIRV to WGSL. [Tint](https://dawn.googlesource.com/tint) is one such tool but maybe [Naga](https://github.com/gfx-rs/naga) is able to do it too.t

**Reply by TalDerei** [reply]
That's really helpful! Do you know what this feature is called in GPU.js by any chance? Thanks @Rezmason and @Kangz for the replies!
